### DISCUSSION TITLE
Collection was modified; enumeration operation may not execute during ViewModel.SaveAsync()

### DISCUSSION THREAD

**Discussion by lwmorris067** [MarimerLLC/csla#4232] [Questions] [answered]
I'm getting this error after calling ViewModel.SaveAsync() from a blazor client. The server-side dataportal works, but then I get this exception in ViewModel.Exception. Any ideas how to troubleshoot this?

DataPortal.Update failed (Exception has been thrown by the target of an invocation.)
Csla.DataPortalException: DataPortal.Update failed (Exception has been thrown by the target of an invocation.)
 ---> Csla.Reflection.CallMethodException: Model.Update method call failed
 ---> System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation.
 ---> **System.InvalidOperationException: Collection was modified; enumeration operation may not execute.**
   at Csla.Core.GraphMerger.MergeBusinessListGraph[T,C](T target, T source)
   at InvokeStub_GraphMerger.MergeBusinessListGraph(Object, Span`1)

**Comment by StefanOssendorf** [comment]
Hi. Do you have async rules which modifies existing collections?

lwmorris067 ***@***.***> schrieb am Do., 26. Sept. 2024,
21:50:

**Reply by lwmorris067** [reply]
No async rules. I do have rules that modify collections.

**Reply by StefanOssendorf** [reply]
Maybe that's the issue here. A BO gets all it's rules executed on arrival at the server. So maybe it's iterating over the list which one of your rules is modifying during that. Hence the collection was modified exception.

**Reply by rockfordlhotka** [reply, marked as answer]
@lwmorris067 do you have any code that interacts with a collection outside of rules? Like constructors or overrides of AddItem?

Constructors and AddItem operations occur during deserialization, possibly PropertyChanged and CollectionChanged events too (I'm not sure) - and if logic is triggered by those operations they could end up running in the middle of the deserialization process.

**Reply by lwmorris067** [reply]
That exactly what's happening. I've added "if (!context.IsPropertyChangedContext) return;" and that seems to have solved the issue, although I have a lot of testing to be absolutely sure.
